### ISSUE TITLE
Realtime LUA sources

### ISSUE THREAD

**Issue by azaz44** [FrSkyRC/ETHOS-Feedback-Community#2431]
We can create sources with `system.registerSource` which is great. But they are not perfect for real time model control, because of how often they are updated. I believe it is better than OpenTX, but I wonder if this could be improved.

I'm aware that LUA takes time to execute, but "source" (or "mixing") scripts are typically short and fast (at least they can be developed as such). And the hardware is improving as well. We have next generation of radios now with X20, X18.

I'm also well aware that using LUA sources for "steering" the model is highly unrecommended because of safety reasons. A script can fail. But still there are situations, where more "real time" response for some channels would be very useful and safe.

So question is, could we have such a possibility, where
- LUA sources would be evaluated every time radio transmits a frame
- LUA sources would have access to current state of inputs / sources (from current frame)

?

----

I'd also appreciate any info on how is current implementation. Is it a defined time where `wakeup` is called every some ms? Is it affected by script execution time, number of scripts or sources or anything?

**Comment by azaz44** [comment]
@bsongis-frsky I wonder if there is a chance of this ever coming? I know this might be not super high priority for many users. But I think this would open a door for so many new possibilities. And I think current wakeup-rate for LUA is not far from that.

**Comment by RealTadango** [comment]
I REALLY need this. I need custom scripts to be running in the background for simple tasks and custom notifications. I really need the custom scripts from OpenTX to be impleneted. Not just to generate inputs but also for other things.

**Comment by bsongis-frsky** [comment]
I don't like running Lua scripts in the Mixer and it is not the case in Ethos, nor in OpenTX (which is slower by the way).
Lua is running in the main task, as it is the case in OpenTX as well. The main loop in Ethos is quicker though because the screen is only refreshed when needed, not always.

**Comment by bsongis-frsky** [comment]
@RealTadango Lua Sources / Lua Tasks should do everything like you do in Mixer Scripts, unless you have a precise counter-example ;)

**Comment by azaz44** [comment]
@bsongis-frsky Is this "closed as completed" because of some change change or just because time response is faster than in Opentx?

What I think is, this would allow for very many possibilities. No matter how powerful Ethos is or will be, it will never be possible to achieve everything with it. For example, we can't do a lot of math with it. And scripts extend possibilities to almost infinity.

This is surely a feature for advanced users, and a feature which you might misuse and crash your model. But this is true for any advanced programming in Ethos or Opentx.

As I understood your post, Ethos has separate "Main" and "Mixer" tasks. And of course stability of "Mixer" task is critical.
But would it be possible to have another task "LUA sources", which would run at same rate as "Mixer"?

I put a longer explanation of that I'm doing below.

------

#### Let me explain what I'm trying to do

I attach the script:
[PolarCyclic.zip](https://github.com/FrSkyRC/ETHOS-Feedback-Community/files/10906481/PolarCyclic.zip)
But it might be complex to analyze because it includes external libs.

But here's the screenshot of its configuration
![image](https://user-images.githubusercontent.com/42830563/223353232-78c1fd0c-eead-494c-ad0c-92c0ffbacc09.png)

Script is meant for helicopters. Here's what it's doing:
- take `Aileron` and `Elevator` stick values as input
- convert them to "polar form". So instead if `Ail` and `Ele` (which one can look at, as cartesian X/Y coordinates) we have `Angle` in which stick is deflected and `Radius` - how much it is deflected
- which greatly matches to what we're dealing with in a helicopter, its disc and its swash plate (which all are circles, which can deflect at any angle)
- now it manipulates these coordinates
  - by applying Expo to `Radius` (instead of individual `Ail` and `Ele` channels)
  - by changing the `Angle`, to effectively rotate the whole steering
  - it also applies `Angle Expo` which is a way of making steering to "stick" to 4 most useful angles (0, 90, 180, 270)
- now it converts it back to Cartesian form (`Ail` and `Ele`), which is what flight controller expects

Some of these calculations might not make a lot of sense for regular gimbals. I modify my gimbals to have smooth response on the cyclic gimbal. So for me Aileron and Elevator become a unified thing, which transfer stick movements directly to swash plate movements of a helicopter. My X20 inside:

![image](https://user-images.githubusercontent.com/42830563/223356144-f594e5cc-ac16-43ba-aad5-5ed0ecc3e872.png)

#### Few words about safety

Here's how safety works, to make sure you can control the model in case the script fails, or has a problem
- The script produces new (modified) `Aileron` and `Elevator` values
- However, this is not what it returns in sources. Rather it returns a difference between new (corrected) value and the old value. So:
  - New source `Ail Offset` is equal to `New_Ail - Old_Ail`
  - New source `Ele Offset` is equal to `New_Ele - Old_Ele`
- In radio, to produce the output for `Aileron` channel there are 2 mixer lines (and same for `Elevator`)
  - 1) use `Aileron` stick as input and redirect to channel number X
  - 2) use `Ail Offset` to *add* to it
- This way we always have main part of `Ail` and `Ele` channels come from Ethos mixers. Only small corrections are added by the script.
- Which allows to protect the system for script problems:
  - If scripts stops working, and its sources are at 0 always, we can still fly the model and land safely. We just loose corrections added by the script.
  - If scripts stops working and its sources freeze - same thing
  - If script - for some reason (like a bug inside) - would produce entirely wrong numbers - we can have logic switches to protect for it. What they could do is to only use `Offset` (correction) mixer lines, when their abs(value) are smaller than some threshold. Corrections are meant to be small.

#### Results

I made this as an experiment at first. The idea seems to be working really good and give very nice feel on controls.

However, because of how LUA sources are updated, it results in channel outputs which are "jumpy" (not smooth)
- Response to sticks is immediate, because the greatest part of the channel output does not depend on LUA
- So after moving a stick somewhere, you see immediate outputs in channels
- and then, some milliseconds later you see a correction (from the script) applied, and channels output jump a bit to a new value
- it is flyable (I test it in flight simulator of course), but is hampered and would probably result in overusing model servos, and also confusing flight controller

Reason for this behavior are, my guess
- wakeup rate of LUA sources (obcious)
- I guess, no sync between LUA cycles and mixer cycles also plays a role

-----

So in reality this is just a problem of making more advanced calculations than Ethos allows, which involve multiple sources.

Another approach would be to have some kind of "Math" module in Ethos, something I was always missing in OpenTX. Take some value, divide by 3, add some other source to it, etc. etc. and produce a value out of that.

But this would be very complex, probably never coming, and I think scripts are a solution for such advanced scenarios.
But response time (and synchronization) are the problem here.

**Comment by azaz44** [comment]
> @RealTadango Lua Sources / Lua Tasks should do everything like you do in Mixer Scripts, unless you have a precise counter-example ;)

@RealTadango You might look at this: https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/2338

This summarizes some problems we have now, and some possible ideas on how to solve them. Key problems are:
- LUA sources in Ethos always must generate a sources (and maybe your script just needs to play a sound or write a file)
- LUA sources don't have any configuration (especially "per model")
- making a script which which outputs multiple sources is somewhat complex

Maybe just join the discussion there to make topic alive :)

**Comment by RealTadango** [comment]
> > @RealTadango Lua Sources / Lua Tasks should do everything like you do in Mixer Scripts, unless you have a precise counter-example ;)
> 
> @RealTadango You might look at this: #2338
> 
> This summarizes some problems we have now, and some possible ideas on how to solve them. Key problems are:
> 
> * LUA sources in Ethos always must generate a sources (and maybe your script just needs to play a sound or write a file)
> * LUA sources don't have any configuration (especially "per model")
> * making a script which which outputs multiple sources is somewhat complex
> 
> Maybe just join the discussion there to make topic alive :)

Yes, that is key funcionality for me also. ETHOS is really lacking this and it should be added to be a good alternative for Open TX / Edge TX.
